packages/block-editor/src/components/media-replace-flow/index.js

```diff
@@ -56,6 +56,17 @@ const MediaReplaceFlow = ( {
 	handleUpload = true,
 } ) => {
 	const [ mediaURLValue, setMediaURLValue ] = useState( mediaURL );
+
+	// In the event of an external change to prop `mediaURL`, synchronize
+	// internal state `mediaURLValue`. Synchronization is especially important
+	// when dealing with images that are being uploaded and whose URL changes
+	// from a blob to a proper URL.
+	useEffect( () => {
+		if ( mediaURL && mediaURL !== mediaURLValue ) {
+			setMediaURLValue( mediaURL );
+		}
+	}, [ mediaURL ] );
```

Comment:
The new React docs recommend [adjusting the state while rendering](https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) for similar cases. But I think this might be easier to understand at a glance.

Reply:
Thanks for the reference! I don't think this was in the recommended practices a couple of years ago, but it makes a lot of sense nowadays, especially in the context of concurrent rendering. :)